File: Haskell/Bits/longueurBit.hs
module LongueurBit (longueurBit) where

import Data.Bits (shiftR)

{-| Renvoie le nombre de bits utilisés dans la représentation binaire d'un nombre entier.

    Input: Int - Le nombre entier dont on veut connaître la longueur de sa représentation binaire.

    Output: Int - Le nombre de bits utilisés.
-}

-- methode itérative sans rappeler la fonction
longueurBit :: Int -> Int
longueurBit 0 = 0
longueurBit n = 1 + longueurBit (shiftR n 1)

-- TEST
main :: IO ()
main = do
    mapM_ (
        (\x -> putStrLn $ "longueurBit " ++ show x ++ " -> " ++ show (longueurBit x)) . (2^))
        [0..9]
{-^ Output:
    longueurBit 1   -> 1
    longueurBit 2   -> 2
    longueurBit 4   -> 3
    longueurBit 8   -> 4
    longueurBit 16  -> 5
    longueurBit 32  -> 6
    longueurBit 64  -> 7
    longueurBit 128 -> 8
    longueurBit 256 -> 9
    longueurBit 512 -> 10
-}

File: Go/Bits/longueurBit.go
package main

import "fmt"

func longueurBit(n int) int {
	nbBit := 0
	for n > 0 {
		nbBit += 1
		n = n >> 1
	}
	return nbBit
}

func main() {
	cases := []int{1, 8, 512, 4096, 16384, 16777216, 35184372088832, 4611686018427387904}
	for _, c := range cases {
		fmt.Printf("%d bits = %d\n", c, longueurBit(c))
	}
}

//Output :
// 1 bits = 1
// 8 bits = 4
// 512 bits = 10
// 4096 bits = 13
// 16384 bits = 15
// 16777216 bits = 25
// 35184372088832 bits = 46
// 4611686018427387904 bits = 63
